File: androidApp/app/src/main/kotlin/com/zeitnot/android/clock/ClockPresenter.kt
package com.zeitnot.android.clock

import com.zeitnot.android.domain.ClockStatus
import com.zeitnot.android.domain.GameServiceContract
import com.zeitnot.android.domain.Player
import com.zeitnot.android.domain.RulesetCategory
import com.zeitnot.android.setup.GameConfiguration

enum class DisplayMode {
    DIGITAL,
    ANALOG
}

data class DialHands(val minuteDegrees: Float, val secondDegrees: Float)

data class ClockFaceModel(
    val side: Player,
    val status: ClockStatus,
    val reading: String,
    val hands: DialHands
)

data class HeaderModel(
    val category: RulesetCategory,
    val baseMinutes: Int,
    val incrementSeconds: Int,
    val moveNumber: Int,
    val isRunning: Boolean
)

data class ControlBarModel(
    val canPause: Boolean,
    val canReset: Boolean
)

data class ClockModel(
    val header: HeaderModel,
    val white: ClockFaceModel,
    val black: ClockFaceModel,
    val controlBar: ControlBarModel,
    val displayMode: DisplayMode,
    val playerToMove: Player,
    val isRunning: Boolean,
    val showPauseDialog: Boolean,
    val showResetDialog: Boolean
)

class ClockPresenter(
    private val configuration: GameConfiguration,
    private val game: GameServiceContract
) : AutoCloseable {

    private var displayMode = DisplayMode.DIGITAL
    private var showResetDialog = false

    fun model(): ClockModel {
        val snapshot = game.snapshot()

        return ClockModel(
            header = HeaderModel(
                category = configuration.category,
                baseMinutes = configuration.baseMinutes,
                incrementSeconds = configuration.incrementSeconds,
                moveNumber = snapshot.moveNumber,
                isRunning = snapshot.isRunning
            ),
            white = faceModel(Player.WHITE),
            black = faceModel(Player.BLACK),
            controlBar = ControlBarModel(
                canPause = snapshot.isRunning,
                canReset = snapshot.isRunning || snapshot.isFinished
            ),
            displayMode = displayMode,
            playerToMove = snapshot.playerToMove,
            isRunning = snapshot.isRunning,
            showPauseDialog = snapshot.isPaused,
            showResetDialog = showResetDialog
        )
    }

    fun press(side: Player) = game.press(side)

    fun pause() = game.pause()

    fun resume() = game.resume()

    fun select(displayMode: DisplayMode) {
        this.displayMode = displayMode
    }

    fun reset() {
        if (game.snapshot().isInProgress) {
            showResetDialog = true
            return
        }

        game.reset()
    }

    fun confirmReset() {
        game.reset()
        showResetDialog = false
    }

    fun cancelReset() {
        showResetDialog = false
    }

    override fun close() = game.close()

    private fun faceModel(player: Player): ClockFaceModel {
        val clock = game.snapshot()[player]

        return ClockFaceModel(
            side = player,
            status = clock.status,
            reading = game.reading(player),
            hands = hands(clock.remainingMillis)
        )
    }

    private fun hands(remainingMillis: Long): DialHands {
        val seconds = remainingMillis / MILLISECONDS_PER_SECOND

        return DialHands(
            minuteDegrees = seconds / SECONDS_PER_HOUR * DEGREES_PER_REVOLUTION,
            secondDegrees = seconds / SECONDS_PER_MINUTE * DEGREES_PER_REVOLUTION
        )
    }

}

private const val MILLISECONDS_PER_SECOND = 1000f
private const val SECONDS_PER_HOUR = 3600f
private const val SECONDS_PER_MINUTE = 60f
private const val DEGREES_PER_REVOLUTION = 360f

File: iosApp/Clock/Sources/Clock/UI/Clock/ClockPresenter.swift
import Observation
import Shared

@Observable
public final class ClockPresenter {

    var displayMode: DisplayMode = .digital
    private(set) var showResetDialog = false

    private let gameConfiguration: GameConfiguration
    private let gameService: any GameServiceProtocol
    private let router: any ClockRoutingProtocol

    public init(
        gameConfiguration: GameConfiguration,
        gameService: any GameServiceProtocol,
        router: any ClockRoutingProtocol
    ) {
        self.gameConfiguration = gameConfiguration
        self.gameService = gameService
        self.router = router
    }

    var headerModel: Header.Model {
        Header.Model(
            category: gameConfiguration.category,
            baseMinutes: timeControl.baseMinutes,
            incrementSeconds: timeControl.incrementSeconds,
            moveNumber: snapshot.moveNumber,
            isRunning: snapshot.isRunning)
    }

    var clocksModel: Clocks.Model {
        Clocks.Model(white: makeClockModel(for: .white), black: makeClockModel(for: .black))
    }

    var controlBarModel: ControlBar.Model {
        ControlBar.Model(canPause: snapshot.isRunning, canReset: snapshot.isRunning || snapshot.isFinished)
    }

    var pauseDialogModel: PauseDialog.Model {
        PauseDialog.Model(playerToMove: dialogPlayer(for: snapshot.playerToMove))
    }

    var showPauseDialog: Bool {
        snapshot.isPaused
    }

    var isCountingDown: Bool {
        snapshot.isRunning
    }

    private var snapshot: GameSnapshot {
        gameService.snapshot
    }

    private var timeControl: TimeControl {
        gameConfiguration.timeControl
    }

    func navigateBack() {
        router.navigateBack()
    }

    func press(_ side: ClockFace.Side) {
        gameService.press(player(for: side))
    }

    func pause() {
        gameService.pause()
    }

    func resume() {
        gameService.resume()
    }

    func reset() {
        guard snapshot.isInProgress else {
            gameService.reset()
            return
        }

        showResetDialog = true
    }

    func confirmReset() {
        gameService.reset()
        showResetDialog = false
    }

    func cancelReset() {
        showResetDialog = false
    }

}

// MARK: Mappers

private extension ClockPresenter {

    func makeClockModel(for player: Player) -> ClockFace.Model {
        ClockFace.Model(
            side: side(for: player),
            state: faceState(for: snapshot[player].status),
            timeDisplay: timeDisplay(for: player))
    }

    func timeDisplay(for player: Player) -> ClockFace.TimeDisplay {
        switch displayMode {
        case .digital: .digital(DigitalFace.Model(reading: snapshot[player].remaining.timeReading))
        case .analog: .analog(AnalogFace.Model(hands: DialHands(remaining: snapshot[player].remaining)))
        }
    }

    func side(for player: Player) -> ClockFace.Side {
        switch player {
        case .white: .white
        case .black: .black
        }
    }

    func player(for side: ClockFace.Side) -> Player {
        switch side {
        case .white: .white
        case .black: .black
        }
    }

    func dialogPlayer(for player: Player) -> PauseDialog.Player {
        switch player {
        case .white: .white
        case .black: .black
        }
    }

    func faceState(for status: ClockStatus) -> ClockFace.State {
        switch status {
        case .awaitingStart: .awaitingStart
        case .toMove: .toMove
        case .lowTime: .lowTime
        case .waiting: .waiting
        case .flagged: .flagged
        }
    }

}
